File: kensingtonbn/kensingtonbn/report/batch_details_with_pricing_rule/batch_details_with_pricing_rule.py
# Copyright (c) 2013, ClefinCode and contributors
# For license information, please see license.txt
import json
import re
from datetime import datetime

import frappe
from frappe import _
from frappe.utils import nowdate, flt
from erpnext.utilities.product import get_price
from erpnext.setup.utils import get_exchange_rate
from erpnext.stock.get_item_details import get_conversion_factor
from erpnext.e_commerce.doctype.e_commerce_settings.e_commerce_settings import get_shopping_cart_settings

def execute(filters=None):
	columns, data = [], []
	data, count = get_data(filters)
	columns = get_columns(filters, count)
	return columns, data

def get_columns(filters, count):
	columns = [
		{
			"fieldname": "chk",
			"label": _(" "),
			"fieldtype": "Data",
		},
		{
			"fieldname": "item",
			"label": _("Item"),
			"fieldtype": "Link",
			"options": "Item",
			"width": 125
		},
		{
			"fieldname": "name",
			"label": _("Batch"),
			"fieldtype": "Link",
			"options": "Batch",
			"width": 125
		},
		{
			"fieldname": "best_value_date",
			"label": _("Best Value Date"),
			"fieldtype": "Date",
			"width": 125
		},
		{
			"fieldname": "total_qty",
			"label": _("Balance Qty"),
			"fieldtype": "Data",
		},
		{
			"fieldname": "buying_rate",
			"label": _("Buying Price"),
			"fieldtype": "Currency",
			"options": "Company:company:default_currency",
			"width": 100
		},
		# {
		# 	"fieldname": "valuation_rate",
		# 	"label": _("Valuation Rate"),
		# 	"fieldtype": "Currency",
		# 	"options": "currency",
		# 	"width": 100
		# },
		{
			"fieldname": "price_list_rate",
			"label": _("Standard Selling Rate"),
			"fieldtype": "Currency",
			"options": "currency",
			"width": 100
		},

	]
	for filter in filters:
		if "warehouse " in filter:
			warehouse, word_to_check = filter, "Kensington"
			if re.search(rf"\b{word_to_check}\b", warehouse, re.IGNORECASE):
				warehouse = re.sub(rf"\b{word_to_check}\b", "", warehouse, flags=re.IGNORECASE)
				warehouse = " ".join(warehouse.split())
			
			columns.extend((
				{
				"fieldname": filter,
				"label": _(warehouse.replace("warehouse", "", 1)),
				"fieldtype": "Data",
				"width": 125
				},
			))
	columns.extend(
		(
		{
			"fieldname": "rate",
			"label": _("Rate"),
			"fieldtype": "Data",
			"width": 100
		},
		{
			"fieldname": "last_modified",
			"label": _("Last Update"),
			"fieldtype": "Date",
			"width": 120
		},
		{
			"fieldname": "btn_add",
			"label": _("Add Pricing Rule"),
			"fieldtype": "Data",
			"width": 120
		},
		{
			"fieldname": "btn_change",
			"label": _("Change Pricing Rule"),
			"fieldtype": "Data",
			"width": 120
		},
		# {
		# 	"fieldname": "recommended_price",
		# 	"label": _("Recommended Price"),
		# 	"fieldtype": "Data",
		# 	"width": 100
		# },
		{
			"fieldname": "confirmed_price",
			"label": _("Confirmed Price"),
			"fieldtype": "Data",
			"width": 100
		},
		{
			"fieldname": "count",
			"label": _("Count"),
			"fieldtype": "Data",
			"hidden": 1,
			"default": count
		}
		
		))
		
	add_warehouses_columns(filters)
	return columns

def get_data(filters):
	batches_data = get_batches_data(filters)
	return batches_data

def get_conditions(filters):
	conditions = ""
	if filters.get("from_date"):
		conditions += " and batch.best_value_date >= '" + filters.get("from_date") + "' "
	if filters.get("to_date"):
		conditions += " and batch.best_value_date <= '" + filters.get("to_date") + "' "
	if filters.get("item"):
		conditions += " and batch.item = '" + filters.get("item") + "' "
	if filters.get("batch") and not filters.get("barcode"):
		conditions += " and batch.name = '" + filters.get("batch") +"' "
	if not filters.get("disabled"):
		conditions += " and batch.disabled = 0 "
	return conditions

def get_batches_data(filters):
	conditions = get_conditions(filters)
	if filters.get("barcode"):
		batch = frappe.db.get_value("Item Barcode", {'barcode': filters.get("barcode")}, 'batch_no')
		conditions += " and batch.name = " + batch
	
	selling_price_list = frappe.db.get_single_value('Selling Settings', 'selling_price_list') or frappe.db.get_value('Price List', _('Standard Selling'))

	strQuery = """
		select CONCAT('<input type="checkbox" class="check_item" data-batch="', batch.name,'" data-item="', batch.item,'">') as chk,
		batch.item,  batch.name, batch.best_value_date,
		IFNULL(
			(SELECT price_list_rate FROM `tabItem Price` WHERE item_code = batch.item and batch_no = batch.name and selling = 1 and price_list = '{price_list}' LIMIT 1),
			(SELECT price_list_rate FROM `tabItem Price` WHERE item_code = batch.item and (batch_no IS NULL or batch_no = '') and selling = 1 and price_list = '{price_list}' LIMIT 1)) 
		AS price_list_rate, 
		IFNULL(
			(SELECT currency FROM `tabItem Price` WHERE item_code = batch.item and batch_no = batch.name and selling = 1 and price_list = '{price_list}' LIMIT 1),
			(SELECT currency FROM `tabItem Price` WHERE item_code = batch.item and (batch_no IS NULL or batch_no = '') and selling = 1 and price_list = '{price_list}' LIMIT 1)) 
		AS selling_currency, 
		CONCAT('<div class="qty" data-batch="', batch.name,'" value="',cast(sum(actual_qty) as float),'">', cast(sum(actual_qty) as float), '</div>') as total_qty,
		{warehouses}
		CONCAT('<a class="btn_pr_add" data-batch="', batch.name,'" data-item="', batch.item,'" style="width:80px; text-align: center; background-color: #e7e7e7; color: black; padding: 5px 30px" onclick="new_price_rule(this)">', 'Add', '</a>') as btn_add,
		CONCAT('<a class="btn_pr_chg" data-batch="', batch.name,'" data-item="', batch.item,'" style="width:80px; text-align: center; background-color: #e7e7e7; color: black; padding: 5px 30px" onclick="change_price_rule(this)">', 'Change', '</a>') as btn_change,
		CONCAT('<input type="number" value="', ROUND(item.confimed_price, 2), '" class="confirmed_price" data-batch="', batch.name,'" data-item="', batch.item,'" style="width:80px; text-align: center; background-color: black; color: white;">') as confirmed_price,
		IFNULL(
			(SELECT rate_of_stock_uom from `tabPurchase Invoice Item` WHERE item_code = batch.item AND batch_no = batch.name and docstatus = 1 ORDER BY creation DESC LIMIT 1),
			(SELECT rate_of_stock_uom from `tabPurchase Invoice Item` WHERE item_code = batch.item AND (batch_no = '' or batch_no IS NULL) and docstatus = 1 ORDER BY creation DESC LIMIT 1)
		) as buying_rate,
		IFNULL(
			(SELECT base_rate from `tabPurchase Invoice Item` WHERE item_code = batch.item AND batch_no = batch.name and docstatus = 1 ORDER BY creation DESC LIMIT 1),
			(SELECT base_rate from `tabPurchase Invoice Item` WHERE item_code = batch.item AND (batch_no = '' or batch_no IS NULL) and docstatus = 1 ORDER BY creation DESC LIMIT 1)
		) as base_rate,
		IFNULL(
			(SELECT conversion_factor from `tabPurchase Invoice Item` WHERE item_code = batch.item AND batch_no = batch.name and docstatus = 1 ORDER BY creation DESC LIMIT 1),
			(SELECT conversion_factor from `tabPurchase Invoice Item` WHERE item_code = batch.item AND (batch_no = '' or batch_no IS NULL) and docstatus = 1 ORDER BY creation DESC LIMIT 1)
		) as conversion_factor
		
		FROM `tabBatch` AS batch 
		LEFT JOIN `tabStock Ledger Entry` AS sle ON batch.name = sle.batch_no
		INNER JOIN `tabItem` AS item ON item.name=batch.item
		WHERE sle.is_cancelled = 0 AND actual_qty!=0
		{conditions}
		GROUP BY batch.name
	""".format(conditions=conditions, 
			warehouses = add_warehouses_columns(filters),
			price_list = selling_price_list)
	
	if filters.get("balance_qty"):
		strQuery += " HAVING sum(actual_qty) > 0 "
	else:
		strQuery += " HAVING sum(actual_qty) >= 0 "
	
	tosort = True
	if filters.get("sort_by") and filters.get("sorting"):
		sort_by = ""
		if filters.get("sort_by") == "Item Code": sort_by = "batch.item"
		elif filters.get("sort_by") == "Batch": sort_by = "batch.name"
		elif filters.get("sort_by") == "Best Value Date": sort_by = "batch.best_value_date"
		else: tosort = False
		if tosort:
			strQuery += f" order by {sort_by} {filters.get('sorting')}"

	start = filters.get("start") - 1
	number_of_rows = filters.get("number_of_rows")

	if filters.get("sort_by", "") != "Last Update":
		strQuery += " LIMIT {0} OFFSET {1} ".format(number_of_rows, start)

	batches = frappe.db.sql(strQuery, as_dict=1)
	shopping_cart_settings = get_shopping_cart_settings()
	customer_group = shopping_cart_settings.default_customer_group
	company = shopping_cart_settings.company 
	company_currency = frappe.db.get_value("Company", company, "default_currency")
	results, exchange_rates = [], {}
	for batch in batches:
		if batch.get('conversion_rate') and batch.get('base_rate', 0) > 0 and batch.get('buying_rate', 0) == 0:
			batch['buying_rate'] = batch.get('base_rate') / batch.get('conversion_factor', 1)
		
		# Setting Prices in Company Currency

		if batch.get('price_list_rate') and batch.get('selling_currency'):
			if not exchange_rates.get(batch['selling_currency']):
				exchange_rate = get_exchange_rate(batch['selling_currency'], company_currency)
				exchange_rates[batch['selling_currency']] = exchange_rate
			else:
				exchange_rate = exchange_rates[batch['selling_currency']]

			batch['price_list_rate'] = batch['price_list_rate'] * exchange_rate

		price = get_price(batch['item'],
			selling_price_list,
			customer_group,
			company,
			1,
			batch['name'], return_pr= True)

		if price:
			# last update : comment these lines to show the rate value based on batch
			# if price.price_list_rate == res[-1][5]:
			# 	price = get_price(batch['item'],
			# 	selling_price_list,
			# 	customer_group,
			# 	company,
			# 	1, return_pr= True)

			if (price.pricing_rule):
				pricing_rule = price.pricing_rule.replace('[','')
				pricing_rule = pricing_rule.replace(']','')
				pricing_rule = pricing_rule.replace('"','')	
				pricing_rule = pricing_rule.strip()
				rate = price.formatted_price_sales_uom
				last_modified = frappe.db.get_value("Pricing Rule", pricing_rule, "modified")
			
			else: 
				pricing_rule = ""
				rate = ""
				last_modified = ""

			price_div = "<div class= 'pricing_rule' data-batch='"+batch['name']+"' data-pricing-rule ='"+pricing_rule+"' value="+str(price.price_list_rate)+">"+rate+"</div>"
			batch["rate"] = price_div
			batch["last_modified"] = str(last_modified)
			
		else: 
			batch["rate"] = 0
			batch["last_modified"] = str(last_modified)

		results.append(batch)
	
	if filters.get("sort_by", "") == "Last Update" and filters.get("sorting"):
		if filters.get("sorting") == "Desc":
			results = sorted(results, key=lambda x: x['last_modified'], reverse=True)
		else:
			results = sorted(results, key=lambda x: x['last_modified'])

		results = results[start: start + int(number_of_rows)]
	return results, len(results)
	
def add_warehouses_columns(filters):
	warehouses = []
	for filter in filters:
		if "warehouse " in filter:
			warehouses.append(filter.replace("warehouse", "", 1))

	warehouses = frappe.db.get_list("Warehouse", filters = {'name': ['in', warehouses], 'disabled': 0} ,fields=['name'], as_list=True)
	strWarehouse = ""
	i = 0
	for wr in warehouses:
		strWarehouse += "SUM(CASE WHEN sle.warehouse = '"+wr[0]+"' THEN actual_qty END) 'warehouse "+wr[0]+"' , "
		i +=1
	return strWarehouse



@frappe.whitelist(allow_guest=True)
def set_pricing():
	batches = frappe.db.get_all("Batch", filters = {'disabled': 0}, fields = ['name', 'item'])

	return batches

@frappe.whitelist()
def add_pricing_rule(item_code, batch, rate):
	rate = flt(rate)
	shopping_cart_settings = get_shopping_cart_settings()

	if rate == 0 or frappe.db.exists("Pricing Rule", {
		"item_code": item_code,
		"batch_no": batch,
		"price_or_product_discount": "Price",
		"company": shopping_cart_settings.company,
		"valid_upto": (">=", nowdate()),
		"currency": frappe.db.get_value("Company", shopping_cart_settings.company, "default_currency"),
		"margin_rate_or_amount": rate,
		"rate_or_discount": "Rate"
	}):
		return

	import random

	doc = frappe.new_doc("Pricing Rule")
	items = {
			"item_code": item_code,
			"batch_no": batch
	}
	doc.update({
		"title": item_code + "_" + str(random.randint(1, 1000)),
		"item_code": item_code,
		"batch_no": batch,
		"price_or_product_discount": "Price",
		"company": shopping_cart_settings.company,
		"valid_from": nowdate(),
		"selling": 1,
		"currency": frappe.db.get_value("Company", shopping_cart_settings.company, "default_currency"),
		"rate": rate,
		"rate_or_discount": "Rate"
	})
	doc.append("items", items)
	doc.insert(ignore_permissions = True)
	return doc.name

@frappe.whitelist()
def update_pricing_rule(rule_name, rate):
	if not frappe.db.exists("Pricing Rule", rule_name):
		return

	doc = frappe.get_doc("Pricing Rule", rule_name)

	doc.update({
		"rate": rate
	})
	doc.save(ignore_permissions = True)
	return doc.name

@frappe.whitelist()
def create_item_label():
	items = frappe.form_dict.get('items', '')
	if items:
		try:
			items = json.loads(items)

		except json.JSONDecodeError:
			frappe.throw("Invalid JSON format received")
	batches = frappe.db.get_all("Batch",{
		"disabled": 0,
		"date_in_report": nowdate(),
	}, "name")

	newitems = []
	if batches:
		for batch in batches:
			bdoc = frappe.get_doc("Batch", batch.name)
			newitems.append({
				"item_code": bdoc.item,
				"item_name": bdoc.item_name,
				"batch_no": bdoc.batch_id,
				"qty": bdoc.qty_in_report,
				"price_list_rate": bdoc.price_in_report
			})

	if newitems and items:
		for item in items:
			found = False
			for newitem in newitems:
				if item["item_code"] == newitem["item_code"]:
					found = True
			if not found:
				newitems.append(item)

	if newitems:
		doc = frappe.new_doc("Item Label")
		doc.update({
			"items": newitems
		})

		return doc
	
@frappe.whitelist()
def set_values(item_code, batch_no, price, qty):
	frappe.db.set_value("Item", item_code, "confimed_price", price)

	doc = frappe.get_doc("Batch", {"batch_id": batch_no}) 
	doc.update({
		"date_in_report": nowdate(),
		"price_in_report": price,
		"qty_in_report": qty
	})
	doc.save(ignore_permissions = True)
	frappe.db.commit()

File: kensingtonbn/kensingtonbn/report/batch_details_with_pricing_rule/batch_details_with_pricing_rule.js
// Copyright (c) 2016, ClefinCode and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Batch Details with Pricing Rule"] ={
	"filters": prepare_filters() ,
	
	onload: function(report) {
		var checked_items = [];
		var recomended_prices  = [];
		var confirmed_prices  = [];
		// setPricing()
		report.page.add_inner_button(__("Add Item Label"), function() {
			const itemsString = JSON.stringify(checked_items);

			fetch('/api/method/kensingtonbn.kensingtonbn.report.batch_details_with_pricing_rule.batch_details_with_pricing_rule.create_item_label', {
				method: 'POST',
				headers: {
					Accept: 'application/json',
					'Content-Type': 'application/json',
					'X-Frappe-CSRF-Token': frappe.csrf_token
				},
				body: JSON.stringify({ "items": itemsString })
			}).then(res => res.json()) // Parse the response to JSON
			.then(r => {
				if (r && r.message){
					var doc = frappe.model.sync(r.message);
					frappe.set_route("Form", doc[0].doctype, doc[0].name);
				}
			})
			.catch(error => console.error('Error:', error));
		
		});
		var style = document.createElement('style');
		
		style.type = 'text/css';
		style.innerHTML = `
			.confirmed_price {
			-moz-appearance: textfield;
			}
			.confirmed_price::-webkit-inner-spin-button,
			.confirmed_price::-webkit-outer-spin-button {
			-webkit-appearance: none;
			margin: 0;
			}
			.round {
				text-decoration: none;
				display: inline-block;
				padding: 8px 16px;
				border-radius: 50%;
				cursor: pointer;
			}
			.previous {
				background-color: #f1f1f1;
				color: black;
			}
			  
			.next {
				background-color: #f1f1f1;
				color: black;
			}
			  
		`;
		document.getElementsByTagName('head')[0].appendChild(style);

		$(document).on('click','input.check_item[type="checkbox"]',function(e) {		
			if ($(this).is(":checked")){
				checked_items.push({
					'item_code': $(this).attr('data-item'),
					'batch_no': $(this).attr('data-batch'),
					'qty': parseInt($('.qty[data-batch = "'+$(this).attr('data-batch')+'"]').attr('value')),
					'price_list_rate': parseFloat($('div.pricing_rule[data-batch = "'+$(this).attr('data-batch')+'"]').attr('value'))
				})
			}
			else {
				for (var i=0; i< checked_items.length; i++){
					if ((checked_items[i]['item_code'] == $(this).attr('data-item')) && 
					(checked_items[i]['batch_no'] == $(this).attr('data-batch'))){
						checked_items.splice(i,1);
						break;
					}
				}
			}
		});
		$(document).delegate('.recommended_price', 'change',  function(e) {	
			var batch = $(this).attr('data-batch');
			var item = $(this).attr('data-item');
			var recommended_price = $('.recommended_price[data-batch = "'+$(this).attr('data-batch')+'"]').val()
			
			if(this.recommended_price_timer){
				clearTimeout(this.recommended_price_timer);
				this.recommended_price_timer= null;
			}

			if(recomended_prices.length > 0){
				for(let i=0;i<recomended_prices.length;i++) {
					if(recomended_prices[i]!= undefined){
						if(recomended_prices[i].item_code === $(this).attr('data-item') && recomended_prices[i].batch_no === $(this).attr('data-batch')){
							delete recomended_prices[i]											
							break
						}
					}										
				}
				recomended_prices.push({
					'item_code': $(this).attr('data-item'),
					'batch_no': $(this).attr('data-batch'),
					'value': recommended_price,
					
				})
			}else{
				recomended_prices.push({
					'item_code': $(this).attr('data-item'),
					'batch_no': $(this).attr('data-batch'),
					'value': recommended_price,
					
				})
			}

			this.recommended_price_timer = setTimeout(() => {											
			frappe.db.set_value('Item', item, 'recommended_price', recommended_price)			
			}, 2000);
					
		
		});
		$(document).delegate('.confirmed_price','click', function(e) {
			$('.dt-row').find('.dt-cell').each(function() {
				this.style.removeProperty("background-color");
			});	

			let parentDiv = $(this).closest('.dt-row');
			
			$(parentDiv).css("background-color", "#B8860B")
			parentDiv.find('.dt-cell').each(function() {
				this.style.setProperty("background-color", "#B8860B", "important");
			});
			$(document).one('click', function(event) {
				// Remove background color if clicked outside
				if (!$(event.target).closest('.dt-row').length) {
					parentDiv.find('.dt-cell').each(function() {
						this.style.removeProperty("background-color");
					});
				}
			});
		
			// Prevent propagation to avoid immediate removal of background color
			e.stopPropagation();
		});		
		
		$(document).delegate('.confirmed_price','change', function(e) {	
			var batch = $(this).attr('data-batch');
			var item = $(this).attr('data-item');
			var confirmed_price = $('.confirmed_price[data-batch = "'+$(this).attr('data-batch')+'"]').val()
			
			if(this.confirmed_price_timer){
				clearTimeout(this.confirmed_price_timer);
				this.confirmed_price_timer = null;
			}

			if(confirmed_prices.length > 0){
				for(let i=0;i<confirmed_prices.length;i++) {
					if(confirmed_prices[i]!= undefined){
						if(confirmed_prices[i].item_code === $(this).attr('data-item') && confirmed_prices[i].batch_no === $(this).attr('data-batch')){
							delete confirmed_prices[i]											
							break
						}
					}										
				}
				confirmed_prices.push({
					'item_code': $(this).attr('data-item'),
					'batch_no': $(this).attr('data-batch'),
					'value': confirmed_price,
					
				})
			}else{
				confirmed_prices.push({
					'item_code': $(this).attr('data-item'),
					'batch_no': $(this).attr('data-batch'),
					'value': confirmed_price,
					
				})
			}
			
			let qty = parseInt($('.qty[data-batch = "'+batch+'"]').attr('value'))
			set_values(item, batch, confirmed_price, qty);
			change_price_rule(this);		
		});
		
		$(document).on('change', '.page-number', function(e) {
			let page = $('.page-number input').val();
			let number_of_rows = frappe.query_report.get_filter_value("number_of_rows")
			frappe.query_report.set_filter_value("start", (number_of_rows * page) - number_of_rows + 1);
			frappe.query_report.refresh();
		})
		$(document).on('click', '.previous', function(e) {
			let page = $('.page-number input').val() || 1;
			if (page > 1){
				let new_number = parseInt(page) - 1
				$('.page-number input').val(new_number).trigger("change");
			}

		})
		$(document).on('click', '.next', function(e) {
			let page = $('.page-number input').val() || 1;
			let new_number = parseInt(page) + 1
			$('.page-number input').val(new_number).trigger("change");
			

		})
		$(document).on('change', 'input[data-fieldname=from_date]', function(e){
			$('.page-number input').val(1)
			frappe.query_report.set_filter_value("start", 1);
			
		})
		$(document).on('change', 'input[data-fieldname=to_date]', function(e){
			$('.page-number input').val(1)
		})
		waitForElm('.dt-scrollable').then((elm) => {
			$(".layout-main-section").append(`
			<div style="text-align: center; margin: 15px;">
				<div class="previous round">&#8249;</div>
				<div class="page-number round"><input type='text' placeholder='1'></div>
				<div class="next round">&#8250;</div>
			</div>
			`)
			$('.page-number').val(1);
			$(elm).scroll(function(){
				for (var i=0; i<checked_items.length; i++){
					var item = checked_items[i]['item_code']
					var batch_no = checked_items[i]['batch_no']
					selector = $('input.check_item[data-item = "'+item+'"][data-batch = "'+batch_no+'"]')
					if (selector.length != 0) {
						if ($('input.check_item[data-item = "'+item+'"][data-batch = "'+batch_no+'"]').is(":checked") == 0) {
							$('input.check_item[data-item = "'+item+'"][data-batch = "'+batch_no+'"]').attr('checked','checked');
						}
					}
				}
				for (var i=0; i<recomended_prices.length; i++){
					if(recomended_prices[i] != undefined){
						var item = recomended_prices[i]['item_code']
						var batch_no = recomended_prices[i]['batch_no']
						var value = recomended_prices[i]['value']
	
						selector = $('.recommended_price[data-item = "'+item+'"][data-batch = "'+batch_no+'"]')
						if (selector.length != 0) {						
							$('.recommended_price[data-item = "'+item+'"][data-batch = "'+batch_no+'"]').attr('value',value);
						}
					}					
				}
				for (var i=0; i<confirmed_prices.length; i++){
					if(confirmed_prices[i] != undefined){
						var item = confirmed_prices[i]['item_code']
						var batch_no = confirmed_prices[i]['batch_no']
						var value = confirmed_prices[i]['value']
	
						selector = $('.confirmed_price[data-item = "'+item+'"][data-batch = "'+batch_no+'"]')
						if (selector.length != 0) {						
							$('.confirmed_price[data-item = "'+item+'"][data-batch = "'+batch_no+'"]').attr('value',value);
						}
					}					
				}

			})
		});
	}
};

function new_price_rule(btn_add){
	var batch = $(btn_add).attr('data-batch');
	var item = $(btn_add).attr('data-item');
	let parentDiv = $(btn_add).closest('.dt-row');
	let priceElement = parentDiv.find('.confirmed_price');
	
	if (priceElement){
		let confirmed_price = priceElement.val()
		if (parseFloat(confirmed_price) == 0){
			frappe.throw("You have to Enter a Price Greater Than Zero!!")
			return
		}
		frappe.call({
			method: "kensingtonbn.kensingtonbn.report.batch_details_with_pricing_rule.batch_details_with_pricing_rule.add_pricing_rule",
			args: {
				"item_code": item,
				"batch": batch,
				"rate": confirmed_price
			},
			freeze: true,
			freeze_message: "Creating a New Pricing Rule...",
			callback: function(r){
				if (r.message){
					const d = new frappe.ui.Dialog({
						'title': 'Updated',
						'fields': [
							{'fieldname': 'ht', 'fieldtype': 'HTML'}
						],
						primary_action_label: 'OK',
						primary_action(values) {
							d.hide();
						}
					});
					
					d.fields_dict.ht.$wrapper.html("A New Pricing Rule is Created. You can Access it from <a href='/app/pricing-rule/"+r.message +"'>here</a>");
					
					$(d.$wrapper).on('keydown', function(e) {
						if (e.which === 13) {
							d.hide();
						}
					});

					d.show();
					}
			}
		})
	}
}

function change_price_rule(btn_chg){
	var batch = $(btn_chg).attr('data-batch');
	var pricing_rule = $('div.pricing_rule[data-batch = "'+batch+'"]').attr('data-pricing-rule');	
	let parentDiv = $(btn_chg).closest('.dt-row');
	let priceElement = parentDiv.find('.confirmed_price');
	if (pricing_rule && priceElement){
		let confirmed_price = priceElement.val()
		if (parseFloat(confirmed_price) == 0){
			frappe.throw("You have to Enter a Price Greater Than Zero!!")
			return
		}
		frappe.call({
			method: "kensingtonbn.kensingtonbn.report.batch_details_with_pricing_rule.batch_details_with_pricing_rule.update_pricing_rule",
			args: {
				"rule_name": pricing_rule,
				"rate": confirmed_price
			},
			freeze: true,
			freeze_message: "Updating an Exist Pricing Rule...",
			callback: function(r){
				if (r.message){
					const d = new frappe.ui.Dialog({
						'title': 'Updated',
						'fields': [
							{'fieldname': 'ht', 'fieldtype': 'HTML'}
						],
						primary_action_label: 'OK',
						primary_action(values) {
							d.hide();
						}
					});
					
					d.fields_dict.ht.$wrapper.html("The Pricing Rule <b><a href='/app/pricing-rule/"+pricing_rule +"'>" + pricing_rule+"</a></b> is updated");
					
					$(d.$wrapper).on('keydown', function(e) {
						if (e.which === 13) {
							d.hide();
						}
					});

					d.show();
					
					
				}
			}
		})
	}
	else if (priceElement){
		new_price_rule(btn_chg)
	}
		
	//frappe.set_route('Form','Pricing Rule', {'name': pricing_rule});
}

function set_values(item, batch, confirmed_price, qty){
	if (parseFloat(confirmed_price) == 0){
		frappe.throw("You have to Enter a Price Greater Than Zero!!")
	}
	frappe.call({
		method: "kensingtonbn.kensingtonbn.report.batch_details_with_pricing_rule.batch_details_with_pricing_rule.set_values",
		args: {
			"item_code": item,
			"batch_no": batch,
			"price": confirmed_price,
			"qty": qty
		}
	})
}
function prepare_filters(){
	filters = [
		{
			"fieldname":"from_date",
			"label": __("From Best Value Date"),
			"fieldtype": "Date",
			"width": "80",
			"default": frappe.datetime.get_today()
		},
		{
			"fieldname":"to_date",
			"label": __("To Best Value Date"),
			"fieldtype": "Date",
			"width": "80",
			"default": frappe.datetime.get_today()
		},
		{
			"fieldname":"item",
			"label": __("Item"),
			"fieldtype": "Link",
			"options": "Item",
			
		},
		{
			"fieldname":"batch",
			"label": __("Batch"),
			"fieldtype": "Link",
			"options": "Batch",
		},
		{
			"fieldname":"barcode",
			"label": __("Barcode"),
			"fieldtype": "Data",
		},
		{
			"fieldname":"disabled",
			"label": __("Show disabled Batches"),
			"fieldtype": "Check",
			"default": 0,
		},
		{
			"fieldname":"start",
			"label": __("Start"),
			"fieldtype": "Data",
			"default": 1,
			"hidden": 1
		},
		{
			"fieldname":"number_of_rows",
			"label": __("Number of Rows"),
			"fieldtype": "Data",
			"default": 12,
		},
		{
			"fieldname":"sort_by",
			"label": __("Sort By"),
			"fieldtype": "Select",
			"options": ['Item Code', 'Batch', 'Best Value Date', 'Last Update'],
			"default": 'Item Code'
		},
		{
			"fieldname":"sorting",
			"label": __("Asc / Des"),
			"fieldtype": "Select",
			"options": ["Asc", "Desc"],
			"default": "Asc",
		},
		{
			"fieldname":"balance_qty",
			"label": __("Balance Qty"),
			"fieldtype": "Check",
			"default": 0,
		},
	]
	frappe.db.get_list("Warehouse",{
		fields: ['name'],
		filters: {'is_group': 0, 'disabled': 0}
	}).then((res)=> {
		for (let i=0; i<res.length; i++){
			let wordToCheck = 'Kensington'
			let regex = new RegExp('\\b' + wordToCheck + '\\b', 'i');
			let cleanedString = res[i]['name']
			if (regex.test(res[i]['name'])) {
				cleanedString = res[i]['name'].replace(new RegExp('\\b' + wordToCheck + '\\b', 'gi'), "");
				cleanedString = cleanedString.replace(/\s+/g, ' ').trim();
			} 
			filters.push(
				{
					"fieldname":"warehouse " + res[i]['name'],
					"label": __(cleanedString),
					"fieldtype": "Check",
					"default": 0,
				},
			)
		}			
	})
	return filters	

}

function waitForElm(selector) {
    return new Promise(resolve => {
        if (document.querySelector(selector)) {
            return resolve(document.querySelector(selector));
        }

        const observer = new MutationObserver(mutations => {
            if (document.querySelector(selector)) {
                resolve(document.querySelector(selector));
                observer.disconnect();
            }
        });

        observer.observe(document.body, {
            childList: true,
            subtree: true
        });
    });
}

